tests/test_contact_specific_functions.py

```diff
@@ -296,10 +424,8 @@ def test_calculate_contact_forces_self_rod():
     )
```

Comment:
Setup looks good to me except that the tangents and lengths need some correction, each tangent should be the unit direction vector between each two consecutive nodes, and the length should be the distance between the consecutive two nodes.

Reply:
Okay thanks, I will change the needful.